File: src/main/java/com/github/svegon/capi/event/network/microsoft_validation/MessageStatusCallback.java
package com.github.svegon.capi.event.network.microsoft_validation;

import net.fabricmc.fabric.api.event.Event;
import net.fabricmc.fabric.api.event.EventFactory;
import net.minecraft.client.network.PlayerListEntry;
import net.minecraft.client.network.message.MessageTrustStatus;
import net.minecraft.network.message.SignedMessage;
import net.minecraft.text.Text;
import org.jetbrains.annotations.Nullable;
import org.spongepowered.asm.mixin.injection.callback.CallbackInfoReturnable;

import java.time.Instant;

public interface MessageStatusCallback {
    Event<MessageStatusCallback> EVENT = EventFactory.createArrayBacked(MessageStatusCallback.class,
            (message, decorated, receptionTimestamp, callback) -> {},
            listeners -> (message, decorated, senderEntry, callback) -> {
        for (MessageStatusCallback listener : listeners) {
            listener.onGetMessageStatus(message, decorated, senderEntry, callback);

            if (callback.isCancelled()) {
                return;
            }
        }
                    });

    void onGetMessageStatus(SignedMessage message, Text decorated, Instant receptionTimestamp,
                            CallbackInfoReturnable<MessageTrustStatus> callback);
}

File: src/main/kotlin/io/github/svegon/mclientapi/event/network/microsoft_validation/MessageStatusCallback.kt
package io.github.svegon.mclientapi.event.network.microsoft_validation

import net.fabricmc.fabric.api.event.Event
import net.fabricmc.fabric.api.event.EventFactory
import net.minecraft.client.network.message.MessageTrustStatus
import net.minecraft.network.message.SignedMessage
import net.minecraft.text.Text
import org.spongepowered.asm.mixin.injection.callback.CallbackInfoReturnable
import java.time.Instant
import java.util.function.Function

fun interface MessageStatusCallback {
    fun onGetMessageStatus(
        message: SignedMessage, decorated: Text, receptionTimestamp: Instant,
        callback: CallbackInfoReturnable<MessageTrustStatus>
    )

    companion object {
        @JvmField
        val EVENT: Event<MessageStatusCallback> = EventFactory.createArrayBacked(
            MessageStatusCallback::class.java,
            MessageStatusCallback { message: SignedMessage, decorated: Text, receptionTimestamp: Instant,
                                    callback: CallbackInfoReturnable<MessageTrustStatus> -> }
        ) { listeners: Array<MessageStatusCallback> ->
            MessageStatusCallback { message: SignedMessage, decorated: Text, senderEntry: Instant,
                                    callback: CallbackInfoReturnable<MessageTrustStatus> ->
                for (listener in listeners) {
                    listener.onGetMessageStatus(message, decorated, senderEntry, callback)

                    if (callback.isCancelled) {
                        return@MessageStatusCallback
                    }
                }
            }
        }
    }
}
